KIPs/kip-247.md

```diff
@@ -0,0 +1,206 @@
+- A2: `GaslessApproveTx.data` is `approve(spender, amount)`.
+- A3: `spender` is a whitelisted `GaslessSwapRouter`.
+- A4: `amount` is a nonzero.
+- A5: `nonce` is `getNonce(tx.from)`.
+
+(Note that above statements can be verified solely through a static validation of a transaction.)
+
+A GaslessApproveTx must meet all the following conditions to be promoted to `txpool.pending`:
+
+- AP1: GaslessApproveTx is followed by a GaslessSwapTx of the same sender and the `GaslessSwapTx` can be promoted to `txpool.pending`.
+
+#### GaslessSwapTx
+
+A `GaslessSwapTx` must meet all the following conditions to be inserted to `txpool.queue`:
+
+- S1: `GaslessSwapTx.to` is a whitelisted `GaslessSwapRouter`.
+- S2: `GaslessSwapTx.data` is `swapForGas(token, amountIn, amountOut, amountRepay)`.
+- S3. `token` is a whitelisted ERC20 token.
+
+(Note that above statements can be verified solely through a static validation of a transaction.)
+
+A `GaslessSwapTx` must meet all the following conditions to be promoted to `txpool.pending`:
+
+- SP1: (If GaslessApproveTx exists in `txpool.queue`) `GaslessApproveTx.to=token`.
+- SP2: (If GaslessApproveTx exists in `txpool.queue`) `GaslessApproveTx.data.amount>=amountIn`.
+- SP3: Nonce is the correct value.
+  - (If GaslessApproveTx exists in `txpool.queue`) `GaslessApproveTx.nonce+1=tx.nonce=getNonce(tx.from)+1`.
+  - (otherwise) `tx.nonce=getNonce(tx.from)`.
+- SP4: `amountRepay` is the correct value.
+  - (If GaslessApproveTx exists in `txpool.queue`) `amountRepay = CalcRepayAmount(GaslessApproveTx, GaslessSwapTx)`.
+  - (otherwise) `amountRepay = amountRepay(GaslessSwapTx)`.
+  - `amountRepay(GaslessApproveTx, GaslessSwapTx) = V1 + V2 + V3` and `amountRepay(GaslessSwapTx) = V1 + V3` where
+    - `V1 = LendTx.Fee() = SwapTx.GasPrice() * 21000`
+    - `V2 = ApproveTx.Fee()` (if exists)
+    - `V3 = SwapTx.Fee()`
+
+While GaslessApproveTx must be followed by with GaslessSwapTx in order to be promoted, GaslessSwapTx can exist by itself.
+
+### Definition of LendTx
+
+Upon detection of a GaslessTx (either GaslessApproveTx or GaslessSwapTx), the block proposer shall inject a transaction, denoted by LendTx, which transfers KAIA to the GaslessTx sender. LendTx must meet all the following requirements:
+
+- L1: `LendTx.type` is 0x7802, a Ethereum dynamic fee transaction type.
+- L2: `LendTx.from` is the proposer itself.
+- L3: `LendTx.to` is `GaslessSwapTx.from`
+- L4: `LendTx.value` is `amountLend`
+  - `amountLend` is `V2 + V3` (from aforementioned `amountRepay` definition).
+
+### GaslessSwapRouter
+
+GaslessSwapRouter is a smart contract responsible for token swap and repayment. It supports only single-hop swaps through a predetermined DEX contract. It must support the following interface:
+
+```
+interface IKIP247 {
+    function swapForGas(address token, uint256 amountIn, uint256 minAmountOut, uint256 amountRepay) external;
+    function addToken(address token, address dex) external;
```

Comment:
Please update to 
```suggestion
    function addToken(address token, address factory, address router) external;
```

Reply:
Updated IKIP247. It includes struct and a few other functions implemented in GaslessSwapRouter.